main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+    while (pending.nonEmpty) {
+      // Extract the next equation and update the pending equations.
+      val e = pending.head
+      pending = pending.tail
+
+      e match {
+        case Equation(Term.Var(x), rhs, _) if !rhs.freeVars.contains(x) =>
+          // Case 1: We have found an equation: `x ~ t` where `x` is not free in `t`.
+          // Construct a singleton substitution `[x -> y]`.
+          val singleton = BoolSubstitution.singleton(x, rhs)
+
+          // Apply the singleton substitution to the unsolved equations.
+          pending = singleton(pending)
+
+          // Apply the singleton substitution to the unsolved equations.
+          unsolved = singleton(unsolved)
+
+          // Compose the new substitution with the current substitution.
+          subst = singleton @@ subst
+
+        case _ => unsolved = e :: unsolved
+      }
+    }
+
+    // Reverse the unsolved equations to ensure they are returned in the original order.
+    (unsolved.reverse, subst)
+  }
+
+  /**
+   * Given a unification equation system `l` computes a most-general unifier for all its equations.
+   *
+   * If multiple equations are involved then we try to solve them in different order to find a small substitution.
+   */
+  private def boolUnifyAllPickSmallest(l: List[Equation]): BoolSubstitution = {
+    // Case 1: We have at most one equation to solve: just solve immediately.
+    if (l.length <= 1) {
+      return boolUnifyAll(l)
+    }
+
+    // Case 2: We have too many equations. We cannot hope to compute with all their permutations: just solve immediately.
+    if (l.length > PermutationLimit) {
+      return boolUnifyAll(l)
+    }
+
+    // Case 3: We have a reasonable number of permutations. We solve all of them and pick the smallest substitution.
+    val results = l.permutations.toList.map {
+      case p => (p, boolUnifyAll(p))
+    }.sortBy {
+      case (_, s) => s.size
+    }
+
+    println("Permutations:")
+    for ((p, s) <- results) {
+      println(s"  $p -- ${s.size}")
+    }
+
+    // Pick the smallest substitution.
+    results.head._2
+  }
+
+  /**
+    * Computes the most-general unifier of all the given equations `l`.
+    *
+    * Throws a [[ConflictException]] if an equation in `l` cannot be solved.
+    *
+    * Note: We assume that any existing substitution has already been applied to the equations in `l`.
+    *
+    * Note: We assume that any existing substitution will be composed with the substitution returned by this function.
+    */
+  private def boolUnifyAll(l: List[Equation]): BoolSubstitution = {
+    var subst = BoolSubstitution.empty
+    var pending = l
+    while (pending.nonEmpty) {
+      val e = pending.head
+      pending = pending.tail
+
+      // Compute the most-general unifier for the current equation `e`.
+      val newSubst = boolUnifyOne(e)
+
+      // Apply the computed substitution to all pending equations.
+      pending = newSubst(pending)
```

Comment:
this function can definitely use a running substitution

Reply:
I added an item to #7363